### PR TITLE
don't allow ? in non-result v2 benchmark function defs, require either Result or () return type

### LINKED ISSUE #13278: benchmarking v2 syntax is confusingly lax about return types, leading to confusion
The benchmarking v2 syntax allows any return type on benchmark function definitions. We should probably only be allowing `()` (blank return type) or `Result<(), BenchmarkError>` in scenarios where the programmer wants to use the `?` operator somewhere in the benchmark function body.

After talking about this with @ggwpez we thought it would make sense to require a return type of `Result<(), BenchmarkError>` in scenarios where the programmer uses the `?` operator within the function body, and also allow a blank `()` return type. Any other return types on the benchmark function definitions would trigger a compiler error. Doing things this way will make the benchmark function defintion feel more like a real function and will make people more aware that they can use `?` in these contexts (and avoid the confusing situation where someone writes `?` and it compiles even though what appears to be the enclosing function def does not have a `Result` or `Option` return type).

Open to any thoughts people may have, though!

### PR DESCRIPTION
**Pull request by sam0x17**
Changes the v2 benchmarking syntax to only accept `()` or `Result<(), BenchmarkError>` as valid return types on benchmark function defs. This change is merely syntactic sugar as the return type is ultimately discarded since the code in these function defs is loaded into a bunch of trait impls. 

- [x] Pallet using this syntax (balances)
- [x] Parsing of valid return types
- [x] Compile error messages for valid return types
- [x] UI tests for valid return types
- [x] Compile error when there is a `?` operator but return type is `()`
- [x] UI test for `?` operator
- [x] use alias to `BenchmarkResult` and simplified parsing
- [x] update UI tests for new simplified parsing
- [ ] Update docs

## Related Issues
* fixes #13278
* #13132

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2334886

File: frame/balances/src/benchmarking.rs [hunk 1]
```diff
@@ -225,7 +225,7 @@ mod benchmarks {
 	}
 
 	#[benchmark]
-	fn force_unreserve() {
+	fn force_unreserve() -> Result<(), BenchmarkError> {
```
Comment (ggwpez) [inline frame/balances/src/benchmarking.rs:228]: We could alias this to `BenchmarkResult`.

File: frame/support/procedural/src/benchmark.rs [hunk 2]
```diff
@@ -152,6 +173,41 @@ struct BenchmarkDef {
...
 	/// Constructs a [`BenchmarkDef`] by traversing an existing [`ItemFn`] node.
 	pub fn from(item_fn: &ItemFn, extra: bool, skip_meta: bool) -> Result<BenchmarkDef> {
+		let invalid_return = |span| {
+			return Err(Error::new(
+				span,
+				"Invalid return type. Only `()` and `Result<(), BenchmarkError>` are allowed.",
+			))
+		};
+		if let ReturnType::Type(_, typ) = &item_fn.sig.output {
+			// parsing for non `()` return type (only allow `Result<(), BenchmarkError>`)
```
Comment (ggwpez) [inline frame/support/procedural/src/benchmark.rs:183]: Could you please put all this added code in its own function? Like `parse_return_type` or whatever.

File: frame/support/procedural/src/benchmark.rs [hunk 3]
```diff
@@ -152,6 +173,41 @@ struct BenchmarkDef {
 impl BenchmarkDef {
 	/// Constructs a [`BenchmarkDef`] by traversing an existing [`ItemFn`] node.
 	pub fn from(item_fn: &ItemFn, extra: bool, skip_meta: bool) -> Result<BenchmarkDef> {
+		let invalid_return = |span| {
+			return Err(Error::new(
+				span,
+				"Invalid return type. Only `()` and `Result<(), BenchmarkError>` are allowed.",
```
Comment (ggwpez) [inline frame/support/procedural/src/benchmark.rs:179]: The `Result<(), BenchmarkError>` is actually a syntax thing and would still work if you alias `Result` to something else, or?

File: frame/support/procedural/src/benchmark.rs [hunk 4]
```diff
@@ -152,6 +173,41 @@ struct BenchmarkDef {
...
+			let _: keywords::BenchmarkError = syn::parse(last_arg.to_token_stream().into())?;
+		} else {
+			// return type is `()` which is permitted
+			// check for question mark ops as these aren't allowed here
+			let mut visitor = ExprTryVisitor::new();
+			visitor.visit_item_fn(item_fn);
+			if let Some(expr) = visitor.results.first() {
+				return Err(Error::new(
+					expr.question_token.span,
+					"The `?` operator can only be used in benchmark functions that specify `Result<(), BenchmarkError>`as their return type"
```
Comment (ggwpez) [inline frame/support/procedural/src/benchmark.rs:206]: Why is this needed? The rust compiler should already mention that the return type is invalid, or?

Comment (sam0x17) [reply frame/support/procedural/src/benchmark.rs:206]: Because in reality you could put any return type on the benchmark function definition and it would compile, even non-sensical ones, because the return type is discarded and replaced by the return type of the traits that get impl'ed. So this whole PR is just making the benchmark function definitions "feel" more like real functions by making the return type follow normal rust semantics to a certain extent

File: frame/support/procedural/src/benchmark.rs (hunk 3 above)
Comment (sam0x17) [reply frame/support/procedural/src/benchmark.rs:179]: yes, I agree an alias would be good to make this look better. I can use that assert impl macro I use elsewhere in this code to assert that whatever is put here is equivalent to `Result<(), BenchmarkError>` I believe

File: frame/support/procedural/src/benchmark.rs (hunk 2 above)
Comment (sam0x17) [reply frame/support/procedural/src/benchmark.rs:183]: yeah I can pull this out.

**Comment by sam0x17** [comment]
So I am going to re-work this a bit to use `static_assertions::assert_impl_all!` and an alias called `BenchmarkResult` that points to `Result<(), BenchmarkError>`, which will also vastly simplify the code and the UX

File: frame/support/procedural/src/benchmark.rs [hunk 5]
```diff
@@ -145,16 +166,38 @@ struct BenchmarkDef {
...
 
+/// Parses the return type of a benchmark function definition
+fn parse_return_type(item_fn: &ItemFn) -> Result<Option<Type>> {
+	match &item_fn.sig.output {
+		ReturnType::Default => {
+			// return type is `()` which is permitted
+			// check for question mark ops as these aren't allowed here
+			let mut visitor = ExprTryVisitor::new();
+			visitor.visit_item_fn(item_fn);
+			if let Some(expr) = visitor.results.first() {
```
Comment (ggwpez) [inline frame/support/procedural/src/benchmark.rs:182]: I think its good to have the syntax clarification, but not sure if this check here actually works.  
It would still be possible to just `return Err("".into())` instead of using `?` and it would exclude `#[block]`s from using it.  
I dont think we can/have to explicitly check this, or?

Comment (sam0x17) [reply frame/support/procedural/src/benchmark.rs:182]: We could have a visitor that checks all return statements and applies a `static_assertions::assert_type_eq_all!` check to each one verifying that it is compatible with `BenchmarkResult`, and that wouldn't be too hard, but honestly now that `BenchmarkResult` is so short I am leaning towards just requiring `-> BenchmarkResult` on _all_ benchmark function defs, which would greatly simplify things and wouldn't be a very heavy lift. Thoughts?

Comment (sam0x17) [reply frame/support/procedural/src/benchmark.rs:182]: Either way would work though I think

**Comment by sam0x17** [comment]
Based on my discussion with @ggwpez I think the best course of action is to actually close this and re-open #13224. There are some clear benefits to generating real functions and one of those benefits is all these weird edge cases will be resolved for free. Additionally some issues, like not using a benchmark parameter within the benchmarking code, will now get warnings that will cause CI to fail (as it should) instead of us having to write crazy visitor patterns and things to identify these ourselves.

So we've come full circle, #13224 is a good idea 😆
